File: kiit-codes/src/commonTest/kotlin/kiit/codes/formats/CodesToProblemTest.kt
package kiit.codes.formats

import kiit.codes.CodesToHttp
import kiit.codes.Err
import kiit.codes.Failed
import kiit.codes.Invalid
import kiit.codes.Restricted
import kiit.codes.StatusConstants
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertNull

class CodesToProblemTest {
    private val catalog = Catalog()
    private val codesToProblem = CodesToProblem(catalog, CodesToHttp())

    @Test
    fun baseUrlDefaultsForKiitOrigin() {
        val status = Failed.Restricted("PAYMENT_REQUIRES_3DS", "3DS required", origin = StatusConstants.KIIT)
        val problem = codesToProblem.build(status)
        assertEquals("https://kiit.dev/problems/restricted/payment-requires-3ds", problem.type)
    }

    @Test
    fun buildThrowsForUnregisteredOrigin() {
        val status = Failed.Restricted("PAYMENT_REQUIRES_3DS", "3DS required", origin = "com.stripe")
        assertFailsWith<IllegalArgumentException> { codesToProblem.build(status) }
    }

    @Test
    fun buildUsesRegisteredBaseUrlAndFullTypePath() {
        catalog.register("com.stripe", "https://stripe.com/problems")
        val status = Failed.Restricted("PAYMENT_REQUIRES_3DS", "3DS required", origin = "com.stripe", scope = "payments.cards")
        val problem = codesToProblem.build(status)
        assertEquals("https://stripe.com/problems/payments.cards/restricted/payment-requires-3ds", problem.type)
    }

    @Test
    fun convertUsesExplicitBaseUrlRegardlessOfCatalog() {
        val problem = codesToProblem.convert(Restricted.DENIED, baseUrl = "https://example.com/probs")
        assertEquals("https://example.com/probs/restricted/denied", problem.type)
    }

    @Test
    fun titleIsStatusMessageAndStatusIsHttpCode() {
        val problem = codesToProblem.build(Restricted.FORBIDDEN)
        assertEquals(Restricted.FORBIDDEN.message, problem.title)
        assertEquals(403, problem.status)
    }

    @Test
    fun noErrLeavesDetailAndInstanceAndErrorsNull() {
        val problem = codesToProblem.build(Restricted.DENIED)
        assertNull(problem.detail)
        assertNull(problem.instance)
        assertNull(problem.errors)
    }

    @Test
    fun errorListPopulatesDefaultErrorDetailEntries() {
        val err =
            Err.ErrorList(
                errors = listOf(Err.ErrorField("email", "not-an-email", "Invalid email")),
                message = "Validation failed",
            )
        val problem = codesToProblem.build(Invalid.INVALID_VALUE, err)
        assertEquals("Validation failed", problem.detail)
        assertEquals(ErrorDetail("email", "Invalid email"), problem.errors?.single())
    }

    @Test
    fun genericBuildMapsThroughACustomErrorItem() {
        data class RichError(override val field: String?, override val message: String, val ref: Any?) : ErrorItem

        val err =
            Err.ErrorList(
                errors = listOf(Err.ErrorField("email", "not-an-email", "Invalid email", ref = "req-42")),
                message = "Validation failed",
            )
        val problem =
            codesToProblem.buildCustom(Invalid.INVALID_VALUE, err) { RichError((it as? Err.ErrorField)?.field, it.message, it.ref) }
        assertEquals("req-42", (problem.errors?.single() as RichError).ref)
    }

    @Test
    fun problemForMatchesBuildWithDefaultErrorDetail() {
        val status = Restricted.FORBIDDEN
        assertEquals(codesToProblem.build(status), problemFor(catalog, CodesToHttp(), status))
    }
}

File: ports/kiit-codes-ts/test/formats/codes-to-problem.test.ts
import { describe, expect, it } from "vitest";
import { Catalog } from "../../src/formats/catalog.js";
import { CodesToProblem, problemFor } from "../../src/formats/codes-to-problem.js";
import { ErrorDetail } from "../../src/formats/error-item.js";
import type { ErrorItem } from "../../src/formats/error-item.js";
import { CodesToHttp } from "../../src/codes.js";
import { ErrorField, ErrorList } from "../../src/err.js";
import { Restricted, Invalid, StatusConstants } from "../../src/groups.js";

// Ported from formats/CodesToProblemTest.kt.

describe("CodesToProblem", () => {
  const catalog = Catalog.of();
  const codesToProblem = CodesToProblem(catalog, CodesToHttp());

  it("baseUrl defaults for the KIIT origin, with code as a query param", () => {
    const status = Restricted("PAYMENT_REQUIRES_3DS", "3DS required", StatusConstants.KIIT);
    const problem = codesToProblem.build(status);
    expect(problem.type).toBe("https://www.kiit.dev/docs/kiit-codes?code=Failed:Restricted:PAYMENT_REQUIRES_3DS#taxonomy");
  });

  it("build throws for an unregistered origin", () => {
    const status = Restricted("PAYMENT_REQUIRES_3DS", "3DS required", "com.stripe");
    expect(() => codesToProblem.build(status)).toThrowError(/No baseUrl registered/);
  });

  it("build uses the registered baseUrl and the full type path", () => {
    const stripeCatalog = Catalog.of({ "com.stripe": "https://stripe.com/problems" });
    const status = Restricted("PAYMENT_REQUIRES_3DS", "3DS required", "com.stripe", "payments.cards");
    const problem = CodesToProblem(stripeCatalog, CodesToHttp()).build(status);
    expect(problem.type).toBe("https://stripe.com/problems/payments.cards/restricted/payment-requires-3ds");
  });

  it("convert uses the explicit baseUrl regardless of the catalog", () => {
    const status = Restricted("PAYMENT_REQUIRES_3DS", "3DS required", "com.stripe");
    const problem = codesToProblem.convert(status, undefined, "https://example.com/probs");
    expect(problem.type).toBe("https://example.com/probs/restricted/payment-requires-3ds");
  });

  it("title is the status message, status is the HTTP code", () => {
    const problem = codesToProblem.build(Restricted.FORBIDDEN);
    expect(problem.title).toBe(Restricted.FORBIDDEN.message);
    expect(problem.status).toBe(403);
  });

  it("no err leaves detail/instance/errors undefined", () => {
    const problem = codesToProblem.build(Restricted.DENIED);
    expect(problem.detail).toBeUndefined();
    expect(problem.instance).toBeUndefined();
    expect(problem.errors).toBeUndefined();
  });

  it("an ErrorList populates default ErrorDetail entries", () => {
    const err = ErrorList([ErrorField("email", "not-an-email", "Invalid email")], "Validation failed");
    const problem = codesToProblem.build(Invalid.INVALID_VALUE, err);
    expect(problem.detail).toBe("Validation failed");
    expect(problem.errors?.[0]).toEqual(ErrorDetail("email", "Invalid email"));
  });

  it("buildCustom maps through a custom ErrorItem", () => {
    interface RichError extends ErrorItem {
      readonly field?: string;
      readonly message: string;
      readonly ref?: unknown;
    }
    const err = ErrorList([ErrorField("email", "not-an-email", "Invalid email", undefined, "req-42")], "Validation failed");
    const problem = codesToProblem.buildCustom<RichError>(Invalid.INVALID_VALUE, err, (e) => ({
      field: e.kind === "ErrorField" ? e.field : undefined,
      message: e.message,
      ref: e.ref,
    }));
    expect(problem.errors?.[0]?.ref).toBe("req-42");
  });

  it("problemFor matches build with the default ErrorDetail", () => {
    const status = Restricted.FORBIDDEN;
    expect(problemFor(catalog, CodesToHttp(), status)).toEqual(codesToProblem.build(status));
  });
});
